src/middle/Compiler.ml

```diff
@@ -1,124 +1,32 @@
-  type syntax_error
-  type frontend_error = (syntax_error, semantic_error) Either.t
-
-  val render_error : frontend_error -> string
-
-  val mir_of_file :
-       opts:frontend_opts
-    -> file:string
-    -> (typed_prog, frontend_error list) result
-
-  val mir_of_string :
-       opts:frontend_opts
-    -> str:string
-    -> (typed_prog, frontend_error list) result
-end
-
-module type Optimize = sig
-  (* variant of possible optimization levels *)
-  type optimization_opts
-
-  (* parse level from string, for use in e.g. command line argument parser *)
-  val optimization_opts_of_string :
-    string -> (optimization_opts, string) result
-
-  val default_optimization_opts : optimization_opts
-  val optimize : opts:optimization_opts -> typed_prog -> typed_prog
-end
-
-module type Backend = sig
-  type backend_opts
-
-  val backend_opts_of_string : string -> (backend_opts, string) result
-  val default_backend_opts : backend_opts
-
-  (* the type of backend representation *)
-  type repr
-
-  val mir_to_repr : opts:backend_opts -> typed_prog -> repr
-  val mir_to_string : opts:backend_opts -> typed_prog -> string
-end
-
-module Compiler = struct
-  module type S = sig
-    type semantic_error
-    type syntax_error
-    type frontend_error
-    type compiler_opts_error
-    type compiler_opts
-
-    val default_compiler_opts : compiler_opts
-
-    val compiler_opts_of_string :
-      string -> (compiler_opts, compiler_opts_error list) result
-
-    val compile_from_file :
-      opts:compiler_opts -> file:string -> (string, frontend_error list) result
-  end
-
-  module Make (F : Frontend) (O : Optimize) (B : Backend) :
-    S
-    with type semantic_error := F.semantic_error
-     and type syntax_error := F.syntax_error
-     and type frontend_error := F.frontend_error = struct
-    type compiler_opts =
-      { frontend_opts: F.frontend_opts
-      ; optimization_opts: O.optimization_opts
-      ; backend_opts: B.backend_opts }
-
-    let default_compiler_opts =
-      { frontend_opts= F.default_frontend_opts
-      ; optimization_opts= O.default_optimization_opts
-      ; backend_opts= B.default_backend_opts }
-
-    [@@@ocaml.warning "-37"]
-
-    type compiler_opts_error =
-      | Frontend_opts_error of string
-      | Optimize_opts_error of string
-      | Backend_opts_error of string
-
-    let compiler_opts_of_string str =
-      Error [Frontend_opts_error ("todo " ^ str)]
-
-    let compile_from_file ~opts ~file =
-      F.mir_of_file ~opts:opts.frontend_opts ~file
-      |> Result.map ~f:(O.optimize ~opts:opts.optimization_opts)
-      |> Result.map ~f:(B.mir_to_string ~opts:opts.backend_opts)
-  end
+include Compiler_intf
+include Core_kernel
+
+module Make
+    (F : Frontend_intf.S)
+    (O : Optimization_intf.S)
+    (B : Backend_intf.S) : S with type frontend_error := F.frontend_error =
+struct
+  type compiler_opts =
+    { frontend_opts: F.frontend_opts
+    ; optimization_opts: O.optimization_opts
```

Comment:
Is this at all related to the PR?

Reply:
No - I tidied it up before I realized I wasn't going to be able to use the signatures yet.